MSC3667: Enforce integer power levels (#3667)
matrix-org/matrix-spec-proposals: proposals/msc3667-enforce-integer-power-levels.md

> # MSC3667: Enforce integer power levels

Reviewer: just to pre-empt some potential discussion: other event types, like join rules, aren't really subject to the same nuances that integer-based power levels are. As a power event itself in state res v2, the power levels are critical for determining whether or not a user can do something - the same issue can theoretically arise from incorrect membership or join rule, however the handling of those two events is targeted at the specific enum values whereas power levels are a range.

> # MSC3667: Enforce integer power levels

Reviewer 2: This MSC has some rather big issues: It implicitly changes the way rooms v1-v9 work, *requiring* implementations to add both string parsers for general stuff and bigints to their implementation for them to be able to claim support for those room versions. Bigints especially are a rather big hurdle to implement in quite some languages. Instead, synapse could fix its behaviour and people could side-upgrade to the same existing room version to fix the rooms properly. Synapse has to follow the spec, not the other way round.

Reviewer 2: Ruma, for example, [has already stated](https://matrix.to/#/!bKOoBjzgyxiMyHIyLp:matrix.org/$dfBXs6oj3FA5YGtZDFQLVbVRcOoYjmrrxdrYygyyIT8?via=matrix.org&via=tedomum.net) that they [do not plan on implementing](https://matrix.to/#/!bKOoBjzgyxiMyHIyLp:matrix.org/$31OVhAm7GlOke2OEPSABzPgIVQpopKZ5cc3hie8hAos?via=matrix.org&via=tedomum.net) bigints, making ruma, and all projects that build ontop of it (e.g. conduit) impossible to support all currently existing room versions.

Reviewer: This MSC does not change the behaviour of v1 through v9 - it's considered a spec bug that those room versions did not previously document it.

Reviewer 2: So then the spec is following synapse, and not the other way round?

Reviewer: We'll be publishing a document which explains how we reach these decisions soon (see https://github.com/matrix-org/matrix-doc/issues/3434 - just waiting on internal review), but the short version is that we have to balance a working federation against historically referenced implementations, and further consider that against the foundation's own values/rules and practicality of any decision.

In this case, Synapse formed the basis for the protocol in the early room versions and the issue has existed well beyond all reasonable expectation for it to be considered a Synapse bug. While unfortunate, it's considered exclusively a spec bug due to the timeline.

----

As an aside, please tone down the accusational language. We're not doing this because we feel like it or because it's fun - we're trying to correct an ancient issue as best we can in the most pragmatic way possible. Not everyone will be happy with the decisions, but that does not excuse bad behaviour.

The github repo is bound by the same code of conduct the community rooms are, and is enforced to the same degree. Everyone is trying to work off the same document and it really sucks when that document fails them - in these cases we aspire to work together to work on a solution and not let frustration of any kind get the better of us.

----

I'm resolving this thread as overly pointed discussion. Further discussion on why/how we reached the point of needing to specify stringy power levels is to be done in #matrix-spec:matrix.org on Matrix, not spread out over a dozen different places.

Reviewer 2: requiring bigints effectively barricated most languages from being able to implement it, having such a spec makes no sense if people can't actually follow it. The spec loses its meaning. Broken rooms will have to upgrade one way or another, be it a /upgraderoom to a new version or the same version, doesn't matter from a UX perspective.

Soru made a comment on the MSC as the issue wasn't raised on the MSC itself yet.

Reviewer 3: I think this discussion is unrelated to this specific MSC, as it only paints the path forward.

#3688 details how to handle the historical aspect, so please comment there; 
https://github.com/matrix-org/matrix-doc/pull/3688#discussion_r796632174

The spec defines power levels in `m.room.power_levels` events as integers, but due to legacy

> behaviour in Synapse, string power levels are also accepted and parsed. The string parsing itself

Reviewer 4: This first sentence is a little confusing. String power levels are not accepted by the spec - but in practice are often accepted by homeserver implementations as not doing so would break existing rooms.

Author: Spec PR #3688 will formally allow today's behaviour into the spec for all room versions up to and including version 9, since this issue was still not yet fixed in Synapse. That will allow us and future implementations to preserve compatibility with existing rooms.

This MSC, on the other hand, will fix the problem in future room versions (hopefully from room version 10).

is problematic because it is done using Python's [int()](https://docs.python.org/3/library/functions.html#int)
function, which has all sorts of associated behaviours:

> > If x is not a number or if base is given, then x must be a string, bytes, or bytearray instance
> > representing an integer literal in radix base. Optionally, the literal can be preceded by + or -
> > (with no space in between) and surrounded by whitespace. A base-n literal consists of the digits 0
> > to n-1, with a to z (or A to Z) having values 10 to 35. The default base is 10. The allowed values
> > are 0 and 2–36. Base-2, -8, and -16 literals can be optionally prefixed with 0b/0B, 0o/0O, or 0x/0X,
> > as with integer literals in code. Base 0 means to interpret exactly as a code literal, so that the
> > actual base is 2, 8, 10, or 16, and so that int('010', 0) is not legal, while int('010') is, as
> > well as int('010', 8).

Reviewer 5: most of this isn't relevant as it requires an explicit base.

```suggestion
> If x is not a number ... then x must be a string, bytes, or bytearray instance
> representing an integer literal. Optionally, the literal can be preceded by + or -
> (with no space in between) and surrounded by whitespace. A base-10 literal consists of the digits 0
> to 9.
```

(notably, all the prefixes here are not valid)

All of this behaviour is exceptionally difficult for non-Python implementations to reproduce

> accurately, so we should not try to do so. 

Reviewer 3: It is not specifically, i've drawn an algorithm to come close; ([original comment](https://github.com/matrix-org/matrix-doc/issues/3264#issuecomment-1026052546))

1. trim all whitespace
2. remove leading `+`
3. if leading character is `-`, remove and remember
4. remove leading zeroes (except the last one, if remaining string is all-zeroes)
5. interpret string as an integer
6. if `-` was encountered, negate the integer

I'm *plenty* sure this is how python does it, and i think at least a mention/help for the spec on how to start to parse the old values would help a ton.

> accurately, so we should not try to do so. 

Reviewer 6: > so we should not try to do so

I think this is a little confusing to add here, as I read it as "current implementations shouldn't do this", rather than "we should avoid doing this in the future".

```suggestion
accurately. 
```

## Proposal

In a future room version, we should enforce the letter of the spec and only allow power levels

> as integers and reject events which try to define them as any other type. This removes all of the

Reviewer 3: My guess would be that it'd fall to canonical JSON integers, as that's the most loose bound in this context.

> as integers and reject events which try to define them as any other type. This removes all of the

Reviewer 7: > My guess would be that it'd fall to canonical JSON integers, as that's the most loose bound in this context.

That would be my interpretation as well, but it doesn't hurt to clarify.

```suggestion
as integers within the range defined by canonical JSON and reject events which try to define them as any other type. This removes all of the
```

> associated headaches with string parsing.

Reviewer: For clarity, which system are we hoping to enforce this within? If it's the auth rules then they might be complicated to represent, but if it's just evaluation-time parsing (ie: non-int is default power level) then that might be a bit easier to write.

Author: I don't think we need to modify the auth rules themselves here. I think an acceptable outcome is that parsing the power level content at the beginning of event auth altogether fails, therefore auth fails with it. This is what the Dendrite implementation does currently. We don't attempt to fall back to a default level, although if anyone has feelings on whether that approach would be better, I'd welcome opinions.

Reviewer: Failing event auth sounds ideal - can we get words to that effect in the MSC? 😇

## Potential issues

We can't avoid the string parsing behaviour altogether because we need to continue to do so for
existing room versions so that we do not break existing rooms. However, there are already documented
cases of this causing problems across implementations. For example, Synapse will accept `"  +2  "` as

> a power level but Dendrite will outright fail to parse it.

Reviewer 4: It would be good to have a line here explicitly stating that: as a result, homeserver implementations may need to update their existing room version implementations to remain compatible with the spec.

Author: It is my hope that the wording in #3688 will be sufficient to cover what implementations should do today and then this MSC can focus on what we do to plug the hole going forward.

Reviewer: (directionally, I think it's best to work on this MSC and that spec PR almost concurrently: the MSC can make reference to what the spec intended to say, as we know we'd accept a clarification about how stringy power levels work, and the spec PR can either continue working on inclusion or wait for this MSC to land and be assigned a room version)

## Alternatives

Event schema enforcement for all event types used in event auth could solve this problem and
more but is a significantly bigger task.

## Unstable prefix

An implementation exists in Dendrite using the `org.matrix.msc3667` room version identifier. The

> experimental room version is based on room version 7, with the additional requirement that power

Reviewer: why room version 7? is there something in v8 or v9 which makes this more complicated to implement?

(concerned as v10 would be based on v9 currently, and this MSC seems to have deliberately chosen a different room version)

Author: I just based it off the most recent version I know to be stable in Dendrite, no other reason. I could easily update it to follow v8 or v9 rules with just a couple lines changed.
